File: articles/feedback-loops-in-resilient-systems/cpp/feedback_delay_simulation.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int main() {
    const int steps = 80;
    const double gain = 0.03;
    const double balancing = 0.14;
    const double target = 75.0;
    const int delay = 5;

    std::vector<double> x(steps + delay + 2, 20.0);

    std::cout << "time,value,target,gain,balancing,delay,overshoot\n";

    for (int t = 1; t < steps; ++t) {
        int delayed_index = std::max(0, t - delay);
        x[t] = x[t - 1] + gain * x[t - 1] - balancing * (x[delayed_index] - target);
        double overshoot = std::max(0.0, x[t] - target);
        std::cout << t << "," << x[t] << "," << target << "," << gain << ","
                  << balancing << "," << delay << "," << overshoot << "\n";
    }

    return 0;
}

File: articles/feedback-loops-in-resilient-systems/fortran/feedback_delay_simulation.f90
program feedback_delay_simulation
  implicit none

  integer, parameter :: steps = 80
  integer, parameter :: delay_steps = 5
  integer :: t, delayed_index
  real :: gain, balancing, target, overshoot
  real, dimension(steps + delay_steps + 2) :: x

  gain = 0.03
  balancing = 0.14
  target = 75.0
  x = 20.0

  print *, "time,value,target,gain,balancing,delay,overshoot"

  do t = 2, steps
     delayed_index = max(1, t - delay_steps)
     x(t) = x(t - 1) + gain * x(t - 1) - balancing * (x(delayed_index) - target)
     overshoot = max(0.0, x(t) - target)

     print '(I0,A,F9.5,A,F9.5,A,F9.5,A,F9.5,A,I0,A,F9.5)', &
        t, ",", x(t), ",", target, ",", gain, ",", balancing, ",", delay_steps, ",", overshoot
  end do
end program feedback_delay_simulation
